Separate out into versioned directories

denzel-morris opened the pull request:
This PR is a WIP, few minor things have to be tweaked and taken care of. Will update this PR and description.

mirons-google commented:
Hey @denzel-morris where do we stand on this PR?

denzel-morris commented:
@mirons-google PR is complete and is update to date with the latest changes in 2.0-forsetisecurity.org-dev. Should be a clean merge.

Note that there are 2 htmlproofer errors for external links that no longer exist that were used in v1.1 of the documentation:

- https://github.com/GoogleCloudPlatform/forseti-security/blob/explain-experimental/scripts/iam_auto_deployment.sh
- https://github.com/GoogleCloudPlatform/forseti-security/tree/explain-experimental

In `_docs/v1.1/quickstarts/explain/index`. The team will have to decide how those links should be updated such they they are not broken anymore.

mirons-google commented on _data/doc_categories.yml:
```diff
@@ -1,14 +1,20 @@
 - title: FAQ
-  href: docs/faq/
+  href: docs/$$VERSION$$/faq/
+v1.1:
```
Will the script create these heading for us?

denzel-morris replied:
Yes, a snapshot of the headings is created such that it pulls whatever is from `latest` and creates a new entry for the new version. So, the latest headings have to be kept inline with the documentation (as was done before).

So, in the case that we have:
```yaml
latest:
- title: <example-title>
  href: <example-url>
v2.0:
- title: Concepts
  href: docs/$$VERSION$$/concepts/
- title: Setup
  href: docs/$$VERSION$$/setup/
- title: Configure
  href: docs/$$VERSION$$/configure/
- title: Use
  href: docs/$$VERSION$$/use/
- title: Develop
  href: docs/$$VERSION$$/develop/
- title: FAQ
  href: docs/$$VERSION$$/faq/
v1.1:
- href: docs/$$VERSION$$/quickstarts/
  title: Quickstarts
- href: docs/$$VERSION$$/howto/
  title: Howtos
- href: docs/$$VERSION$$/guides/
  title: Guides
```
when a new version is generated for say `v2.1` it will become:
```yaml
latest:
- title: <example-title>
  href: <example-url>
v2.1:
- title: <example-title>
  href: <example-url>
v2.0:
- title: Concepts
  href: docs/$$VERSION$$/concepts/
- title: Setup
  href: docs/$$VERSION$$/setup/
- title: Configure
  href: docs/$$VERSION$$/configure/
- title: Use
  href: docs/$$VERSION$$/use/
- title: Develop
  href: docs/$$VERSION$$/develop/
- title: FAQ
  href: docs/$$VERSION$$/faq/
v1.1:
- href: docs/$$VERSION$$/quickstarts/
  title: Quickstarts
- href: docs/$$VERSION$$/howto/
  title: Howtos
- href: docs/$$VERSION$$/guides/
  title: Guides
```

denzel-morris replied:
It should be noted, that if the categories are changed then they must be updated in the `_config.yml` as well such that the defaults are set properly. When creating a new version, the `./scripts/create_new_version_from_latest.sh` pulls the version config defaults from `./scripts/data/version_config_defaults.yml` and adds them to `_config.yml`.

So in the example above, where the categories are modified in latest, they should also be modified in `./scripts/data/version_config_defaults.yml` to look like:
```yaml
  - scope:
      path: _docs/$$VERSION$$
    values:
      category: Overview ($$VERSION$$)
      layout: docs
  - scope:
      path: <example-url>
    values:
      category: <example-title> ($$VERSION$$)
```
This is as it was before. The `_config.yml` and `_data/doc_categories.yml` have to be kept in-sync for categories to work properly. There's no easy way to automated detection of new categories because the titles and url mappings have to be written by the documentation writers -- or more specifically whomever is responsible for the information architecture.

mirons-google replied:
Thanks. Can we just be sure this is part of the developer docs? Also can we be sure to consistently added a comment to the head of every file that it's auto-generated?

denzel-morris replied:
Absolutely, and sure thing. Just a note re: `_data/doc_categories.yml`, while a portion of it is auto-generated, the `latest:` key is meant to be maintained by documentation authors. Therefore it's a "auto-generate" but "do modify", therefore, I will note as such in the file head.

mirons-google commented on _data/doc_categories.yml.v1.1:
```diff
@@ -0,0 +1,10 @@
+# The order of the categories in this file (first-to-last) determines their
```
Is this also auto-generated?

denzel-morris replied:
Nope, this file was included mistakenly and is removed. Sorry about that.

mirons-google commented on _includes/site/layout/docs/tabs.html:
```diff
@@ -1,23 +1,46 @@
-    </a>
+  <div class="col-md-4 text-right">
+    <div class="btn-group" id="version-dropdown" style="margin-top: 1px; margin-bottom: 0; margin-right: 30px">
+      <a data-target="#version-dropdown" class="btn btn-raised dropdown-toggle" data-toggle="dropdown" aria-expanded="false">
```
the class "btn-raised" is having weird issues where the text it in goes away. Can we make this btn-secondary?

denzel-morris replied:
Sure thing, may I ask what browser and resolution you're experiencing this at? I cannot reproduce. Change to `btn-secondary` nonetheless.

mirons-google replied:
I followed-up off PR.

mirons-google commented on _includes/site/layout/docs/tabs.html:
```diff
@@ -1,23 +1,46 @@
+        <span class="caret"></span>
+      </a>
+      <ul class="dropdown-menu">
+        {% assign doc_name = page.path | split:"/" | slice:2,999 | join:"/" %}
```
Can you add a comment here as to what this liquid logic is doing? and Why the choice of the magic numbers 2,999 and 3,999?

denzel-morris replied:
Documented in forthcoming update.

mirons-google replied:
Thanks. Can I ask that in addition to overall docs can you add something here in the code too?

mirons-google commented on _includes/site/layout/docs/tabs.html:
```diff
@@ -1,23 +1,46 @@
+    </div>
   </div>
-</div>
+</div>
```
Final newline please.

denzel-morris replied:
Newlines addressed in forthcoming update.

mirons-google commented on _plugins/exclude_from_watch.rb:
```diff
@@ -0,0 +1,12 @@
+      Array(options['exclude'] + options['exclude_from_watch']).map { |e| Jekyll.sanitized_path(options['source'], e) }
+    end
+  end
+end
```
Final newline please.

mirons-google commented on _plugins/autobuild_latest.rb:
```diff
@@ -0,0 +1,9 @@
+
+Jekyll::Hooks.register :site, :post_write do |site|
+  `rm -rf _docs/latest/`
+end
```
Final newline please.

mirons-google commented on _plugins/exclude_from_watch.rb:
```diff
@@ -0,0 +1,12 @@
+require 'jekyll-watch'
```
Can you add some docs here?

mirons-google commented on _plugins/autobuild_latest.rb:
```diff
@@ -0,0 +1,9 @@
+require 'jekyll'
```
Can you add some docs here?

denzel-morris replied:
Docs added for plugins.

mirons-google commented on scripts/build_doc_version_index.sh:
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
```
Can you follow this [style guide](https://google.github.io/styleguide/shell.xml) for your bash scripts?

We're also indenting 4 spaces, thanks.

denzel-morris replied:
Just to make confirm, we're indenting 4 spaces as opposed to the 2 spaces [recommended](https://google.github.io/styleguide/shell.xml?showone=Indentation#Indentation) by the style guide?

mirons-google replied:
Yes please, thanks.

denzel-morris replied:
Haha, no worries, I've updated everything else to adhere as best as possible to style guide.

mirons-google commented on scripts/build_doc_version_index.sh:
```diff
@@ -0,0 +1,77 @@
+    filename="${file#./}"
+    printf "$filename:\n" >> _data/doc_version_index.yml
+
+    # for version_directory in $version_directories; do
```
If this code isn't needed (commented out) go ahead and remove it please. Thanks.

denzel-morris replied:
Removed.

mirons-google commented on scripts/generate_tombstone_files.sh:
```diff
@@ -0,0 +1,51 @@
+  done
+}
+
+generate_tombstone_files
```
Final newline.

mirons-google commented on travis/jekyll_build.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Please also add the License to any new files that aren't documentation. Fix throughout. Thanks.

denzel-morris replied:
I'm sorry, I'm not understanding this: specifically, what license and where? Only thing I can think of is `_plugins/` but since that's all custom code built specifically for us, it should share the license of the repository.

mirons-google replied:
Apologies on being vague. Where we can we'll need to add our Apache 2.0 License.

https://github.com/GoogleCloudPlatform/forseti-security/blob/dev/setup.py#L1

mirons-google replied:
You'll also need to chmod 755 all of these shell scripts otherwise Travis won't be able to run them.

denzel-morris commented:
@mirons-google I've responded to all your feedback, please let me know of anything additionally. Also, the workflow re: versioning will be documented more fully in the README in a subsequent PR as well.

mirons-google commented on travis/jekyll_build.sh:
```diff
@@ -0,0 +1,13 @@
+if  [ -z ${JGT+x} ]; then
+    # We don't have access to the encrypted vars.
+    bundle exec jekyll build
+else:
```
In testing, this colon should go away

denzel-morris commented:
@mirons-google I appreciate the review; updated per your feedback. Added Apache license to all scripts and plugins; added comment to `doc_categories.yml` re: auto-generation and keeping `latest` key up-to-date with documentation; chmod 755 on `scripts`.

mirons-google commented on scripts/create_new_version_from_latest.sh:
```diff
@@ -153,7 +171,8 @@ function main() {
     update_docs_links
 
-    echo "Snapshotting latest includes to '_includes/docs/${RELEASE_VERSION}' ..."
+    echo "Snapshotting latest includes to '_includes/docs/${RELEASE_VERSION}'
```
nit: please use the "\" line continuation

mirons-google commented:
Thanks @denzel-morris it looks like you have some conflicts with the base branch. Can you merge for me before I approve? Thanks.

mirons-google commented:
@denzel-morris I just pulled this forked branch and it seems the `jekyll_build.sh` script doesn't have the +x bit set.

mirons-google commented on travis/jekyll_build.sh:
```diff
@@ -3,7 +3,7 @@
 if  [ -z ${JGT+x} ]; then
```
nit: I think this file needs a LICENSE too, thanks.
